Dockerfile

```diff
@@ -0,0 +1,60 @@
+FROM ubuntu:22.04
+
+# Update the package manager
+RUN apt-get update
+
+# Install OpenJDK 17
+RUN apt-get install -y openjdk-17-jdk
+
+# Install wget and curl
+RUN apt-get install -y wget curl
+
+# Install Leiningen 2.9.1
+RUN wget https://raw.githubusercontent.com/technomancy/leiningen/2.9.1/bin/lein \
+    && mv lein /usr/local/bin/ \
+    && chmod +x /usr/local/bin/lein \
+    && lein
+
+# Install Clojure
+RUN wget https://download.clojure.org/install/linux-install-1.10.3.986.sh \
+    && chmod +x linux-install-1.10.3.986.sh \
+    && ./linux-install-1.10.3.986.sh
+
+# Install Node.js LTS and npm
+RUN apt-get install -y curl \
+    && curl -sL https://deb.nodesource.com/setup_lts.x | bash - \
+    && apt-get install -y nodejs \
+    && npm install npm@latest -g
+
+RUN npm install -g sass@1.51.0
+
+# Create a directory for the application
+RUN mkdir /app
+
+# Copy project.clj into the Docker image to make it cachable
+COPY project.clj /app/
+
+# Set the working directory to /app
+WORKDIR /app
+
+# Install project dependencies using lein
+RUN XTDB_ENABLE_BYTEUTILS_SHA1=true lein deps
+
+# Copy the current directory into the container
+COPY . /app
```

Comment:
As I mentioned in Discord, I'm not sure what this command is doing, given the `WORKDIR /app` command from above. Assuming this `COPY` still makes sense, the rest of this PR lgtm.

Reply:
WORKDIR doesn't copy the content of the current directory. It just changes the default directory where the subsequent container image commands will run from. Saves us from having to do a `cd app` each time.

Reply:
I'm still confused, sorry. If you `WORKDIR /app`, is the current directory not then `/app`? Maybe a better way of asking this is whether these two are equivalent, in isolation:

```
WORKDIR /app
COPY . /app
```

vs

```
cd /app
COPY . /app
```

...both of these look like the `COPY` is just copying a directory into itself, but I think I'm probably misunderstanding `WORKDIR`. If I am, maybe the comment above the `COPY` command could be made clearer for folks like me who aren't familiar with Docker? :) Thanks and sorry for the (my) confusion!

Reply:
Think of every line in dockerfile as something that runs in as a command in a child shell that gets spawned from the first shell.

Let's say my default work directory is `~` and I run two commands:

```Dockerfile
RUN cd /app
RUN pwd
```
the output that I shall see for pwd would be the home directory, e.g. /home/ubuntu

because in a way you can see this as one time shell command `sh -c 'cd /app'` the moment that run command finishes it has no memory of the change of the directory.

WORKDIR is a dockerfile keyword that changes the default for every subsequent command. 

I am guessing the source of confusion could be because there are implementations where the local directory is mounted as a directory within the docker container. We are not doing that but copying over the code inside the container image and letting it be a standalone entity.

Hence COPY is needed. Because without it /app doesn't have any code.

Reply:
Also there is a bit of tribal knowledge of how to make the image layers more cacheable. I have added more comments to the dockerfile that explains this.

what we would like to faster rebuild of image locally when testing. To be able to do it, the dynamic content should be copied as late as possible.

we could also have chosen
```Dockerfile
# Copy the current directory into the container
COPY . /app

RUN XTDB_ENABLE_BYTEUTILS_SHA1=true lein deps

```
But then this would essentially invalidate the cache each time I make even a single line change in my local codebase. And it would increase the image rebuild times from what my testing by roughly 4 minutes 🙂

Reply:
Gotcha! Thanks for this explanation, @balwa ... it wouldn't have been necessary if I knew anything at all about Docker, obviously. ;)